### PR TITLE
feat(apigee-skills-serving): add skill-finder + skill-publisher + one-line provisioning

### PR DESCRIPTION
**Pull request by gsjurseth**
## Summary

Extends the `references/apigee-skills-serving/` reference (merged in
#889) with two shipped agent skills and a one-line onboarding
experience.

**New skills:**
- **`skills/skill-finder/`** — consumer-side discovery+install client.
  Queries the customer's API hub for signed skills, verifies their
  Ed25519 signatures against locally-installed per-deployment trust
  roots, and installs into the runtime skills directory (Gemini CLI,
  OpenCode, Antigravity).
- **`skills/skill-publisher/`** — author-side pipeline orchestrator.
  Wraps pack → sign → upload → register into a single skill an agent
  can invoke on a source skill directory.

**New bin/ scripts:**
- **`bin/install-skill-finder.sh`** — one-line curl installer for
  skill-finder + skill-publisher into any of three agent runtimes.
- **`bin/install-skill-publisher.sh`** — same, publisher-only.
- **`bin/provision.sh`** — end-to-end customer bootstrap. Enables
  GCP APIs, creates the GCS bucket, sets up the API hub attribute
  taxonomy, generates a fresh Ed25519 signing keypair on the
  operator's machine, publishes the three example skills, and drops
  the trust root into skill-finder's `keys/`.

The five-minute path for a new customer:

```bash
# 1. Install skill-finder + skill-publisher on your machine.
curl -fsSL https://raw.githubusercontent.com/apigee/devrel/main/\
references/apigee-skills-serving/bin/install-skill-finder.sh \
  | bash -s -- --runtime gemini

# 2. Clone the reference and provision your project.
git clone https://github.com/apigee/devrel.git
cd devrel/references/apigee-skills-serving
bash bin/provision.sh --project MY_PROJECT --yes
```

## Trust model (deliberate design)

The `apigee/devrel` repo intentionally contains **no signing key
material and no signed release bundles**. This constraint drove
several design choices:

- **skill-finder itself is trusted by provenance**, not by
  signature. TLS + the `apigee` GitHub org is the trust boundary.
  This matches how `kubectl`, `gh`, and `terraform` installers
  work.
- **Payload skills are trusted by signature** against
  **per-deployment trust roots**. Each customer's `provision.sh`
  generates a fresh Ed25519 keypair on their machine; the private
  key never leaves the machine and is never committed to any
  repo.
- **Multi-key trust roots** (`keys/*.pem`) support multi-publisher
  orgs and zero-downtime key rotation. A downloaded manifest is
  accepted if its `signing_key_id` matches ANY installed key.

Full trust-model rationale in
[`docs/architecture.md`](https://github.com/apigee/devrel/blob/main/references/apigee-skills-serving/docs/architecture.md#trust-model).
Customer runbook in
[`docs/provisioning.md`](https://github.com/apigee/devrel/blob/main/references/apigee-skills-serving/docs/provisioning.md).

## Explicit non-goals

- **No signed release bundles attached to `apigee/devrel` releases.**
  release-please's monorepo cadence is unchanged.
- **No signing key in GitHub Actions.** Zero signing infrastructure
  in this repo. Every signature happens on the customer's machine.
- **No changes to the existing publisher scripts** (`pack_skill`,
  `sign_skill`, `upload_skill`, `register_skill`,
  `update_taxonomy`). This PR only adds new consumers of them.
- **No taxonomy additions beyond `skill-spec`** (which `provision.sh`
  patches into `system-spec-type`; the same PATCH the demo scripts
  have needed since #889). v2 trust-root-in-API-hub design is
  documented but deferred.

## Change breakdown

7 commits totalling ~5389 LoC across 20 files:

| Commit    | Type | Files | Description                                                                     |
|:----------|:-----|:------|:--------------------------------------------------------------------------------|
| `3a02054d`| fix  | 1     | `demo-setup.sh` env-ordering bug (small carry-over: pre-flight was called before env export) |
| `babed2e6`| feat | 9     | Copies `skill-finder` + `skill-publisher` source in; multi-key trust-root code changes in `find_install.py` |
| `df3ae710`| feat | 2     | `bin/install-skill-finder.sh` + `bin/install-skill-publisher.sh` — the one-line installers |
| `e0cbc4fa`| feat | 1     | `bin/provision.sh` — the customer bootstrap script (idempotent across all 7 steps) |
| `2e59ab3f`| docs | 4     | New `docs/provisioning.md`; extends `docs/architecture.md` trust-model section; updates `README.md` + `docs/publish-and-install.md` |
| `362c97a5`| fix  | 2     | Bandit + flake8 fixes surfaced by local MegaLinter run                          |
| `1b4d71f6`| test | 1     | 17 new unit tests for the multi-key `_load_pubkeys` + `_verify_signature` functions |

## Verification (all run locally before this PR)

| Loop                                                                                                        | Result             |
|:------------------------------------------------------------------------------------------------------------|:-------------------|
| shellcheck on all four new bash scripts                                                                     | PASS               |
| flake8 on `skills/skill-finder/scripts/`                                                                    | PASS (0 errors)    |
| pytest full suite                                                                                           | 237/237 in 1.4s    |
| pytest full suite as `uid=0(root)` inside `python:3.13-slim` (matches Cloud Build env)                      | 237/237            |
| `pipeline.sh` local run (Cloud Build equivalent)                                                            | 237/237            |
| MegaLinter v4 via podman (same env vars as `.github/workflows/devrel-static-checks.yml`)                    | `◬` (only non-blocking) |
| `bin/install-skill-finder.sh --repo gsjurseth/devrel --ref feat/add-skill-finder` — real fetch + install    | PASS               |
| `bin/provision.sh --check-only` against a fully-provisioned project (7/7 steps report OK)                   | PASS               |
| `bin/provision.sh` wet re-run against fully-provisioned project (idempotent, zero writes, exits at READY)   | PASS               |
| End-to-end: skill-finder installed via installer, populated with a trust root, resolves + installs a payload skill signed with the matching key | PASS (16-line contract) |

## Two design decisions worth flagging in review

1. **`GCS bucket = public-read`** (`allUsers:objectViewer`) is a
   demo-only choice. skill-finder currently downloads via anonymous
   HTTPS (`https://storage.googleapis.com/<bucket>/<file>.skill`),
   no ADC scope. Production would use signed URLs or an
   authenticated consumer service account; adding that would be a
   ~50 LoC change to `find_install.py._download_zip`. Called out
   in `docs/provisioning.md#step-3-gcs-bucket` and README
   Limitations.

2. **`skill-spec` enum patch to `system-spec-type`** is done via
   direct API hub `PATCH` from `provision.sh` because
   `scripts/update_taxonomy.py` only creates user-defined
   attributes (not system-defined). Called out as a known gap in
   `docs/provisioning.md#step-4-api-hub-attribute-taxonomy`; a
   future PR could migrate the patch into `update_taxonomy.py`.

## Follow-ups (out of scope for this PR)

- **`docs/releasing.md`** — deliberately not added. The reference
  implementation ships with the devrel monorepo's release-please
  cadence; installers fetch `main` by default. No separate release
  channel to document.
- **v2 API-hub-discovered trust root** — documented in
  `architecture.md#trust-root-discovery-in-v2` but not
  implemented. Would eliminate the "operator manually populates
  keys/" step by having skill-finder fetch its trust root from
  API hub itself.
- **Antigravity install path verification** — `install-skill-finder.sh`
  supports `--runtime antigravity` (installs to
  `~/.gemini/antigravity/skills/`) but I have not tested this
  live (no antigravity runtime available in the dev environment).
  The code path is shared with the Gemini CLI install (tested
  green), so risk is low, but flag if a maintainer has an
  antigravity setup.

### CODE REVIEW

File: references/apigee-skills-serving/bin/install-skill-finder.sh [hunk 1]
```diff
@@ -0,0 +1,546 @@
...
+  fi
+
+  if [ "$DRY_RUN" -eq 1 ]; then
+    log "DRY RUN: skipping copy + venv-wrapper install"
+    return
+  fi
+
+  # Atomic install: build in a staging dir, mv into place.
+  local staging
+  staging=$(mktemp -d -p "$INSTALL_ROOT" ".staging-XXXXXX")
```
Comment (OmidTahouri) [inline references/apigee-skills-serving/bin/install-skill-finder.sh:440]: `$INSTALL_ROOT` may not exist yet on a fresh system when `mktemp -d -p "$INSTALL_ROOT"` is called here (the `mkdir -p "$INSTALL_ROOT"` currently occurs later at line 502).

```suggestion
  mkdir -p "$INSTALL_ROOT"
  local staging
  staging=$(mktemp -d -p "$INSTALL_ROOT" ".staging-XXXXXX")
```

File: references/apigee-skills-serving/bin/install-skill-finder.sh [hunk 2]
```diff
@@ -0,0 +1,546 @@
...
+log "  1. Run bin/provision.sh from the apigee/devrel checkout"
+log "     to generate a fresh ed25519 keypair, publish the initial"
+log "     example skills to your API hub signed with the new key,"
+log "     and drop the corresponding public key into keys/."
+log ""
+log "  2. Drop your org's existing public key PEM into"
+log "     $INSTALL_ROOT/skill-finder/keys/<fingerprint>.pem"
+log "     manually. Any file matching keys/*.pem is loaded."
+log ""
+log "See docs/trust-root.md for the full trust-model discussion."
```
Comment (OmidTahouri) [inline references/apigee-skills-serving/bin/install-skill-finder.sh]: `docs/trust-root.md` does not exist in the repository (the trust model is documented in `docs/architecture.md#trust-model`).

```suggestion
log "See docs/architecture.md#trust-model for the full trust-model discussion."
```

File: references/apigee-skills-serving/skills/skill-publisher/scripts/publish.sh [hunk 3]
```diff
@@ -0,0 +1,266 @@
...
+log()  { echo "[skill-publisher] $*"; }
+err()  { echo "[skill-publisher] $*" >&2; }
+
+# ---------------------------------------------------------------
+# Pre-flight: required flags present and non-empty
+# ---------------------------------------------------------------
+for var in SRC BUCKET PRIV_KEY PROJECT LOCATION; do
+  if [ -z "${!var}" ]; then
+    err "FATAL: --${var,,} is required (got empty)"
+    exit 1
```
Comment (OmidTahouri) [inline references/apigee-skills-serving/skills/skill-publisher/scripts/publish.sh:101]: `${var,,}` parameter expansion is a Bash 4.0+ feature and causes a syntax error (`bad substitution`) under macOS default `/bin/bash` (Bash 3.2), which conflicts with the portability goal stated in line 57.

```suggestion
    flag_name="$(echo "$var" | tr '[:upper:]' '[:lower:]' | tr '_' '-')"
    err "FATAL: --${flag_name} is required (got empty)"
```

**Review by OmidTahouri — COMMENTED** [review]
Thank you for putting this together! The overall architecture, multi-key trust root model, and hermetic test suite are great additions to `references/apigee-skills-serving/`.

Overall this looks solid. There are a few portability edge cases (macOS Bash 3.2, fresh install directory initialization), an Antigravity path discrepancy between the installer and `find_install.py`, and minor documentation references to clean up before merge:

* **Taxonomy table in `docs/architecture.md`**: Note that the table in section _"API hub attribute taxonomy"_ still lists the earlier attribute names (`skill-compatible`, `skill-runtime-iam`, etc.) rather than the active schema (`agentic_skill`, `keywords`, `signing_key_id`, `gs_uri`, and `system-spec-type` with `skill-spec`). Updating that table to match the implementation would be helpful.

Please see the in-line comments for details and suggestions on the other items.

File: references/apigee-skills-serving/skills/skill-finder/scripts/find_install.py [hunk 4]
```diff
@@ -0,0 +1,1322 @@
...
+SKILL_DIR = Path(__file__).resolve().parent.parent
+# Multi-key trust root: every PEM under keys/ is a trusted
+# publisher. A manifest is accepted if its signing_key_id
+# fingerprint matches ANY installed key. See keys/README.md for
+# the operational model (provision.sh drops the key, operators
+# can add or remove keys manually).
+KEYS_DIR = SKILL_DIR / "keys"
+
+
+_ANTIGRAVITY_ROOT = Path.home() / ".gemini" / "config" / "skills"
```
Comment (OmidTahouri) [inline references/apigee-skills-serving/skills/skill-finder/scripts/find_install.py]: There is an Antigravity path mismatch: `install-skill-finder.sh` (line 193), `provision.sh` (lines 711, 741), and `SKILL.md` (line 74) use `~/.gemini/antigravity/skills/`, whereas `find_install.py` checks against `~/.gemini/config/skills/`.

When installed via `install-skill-finder.sh --runtime antigravity`, `_detect_skills_root()` fails to match and falls back to OpenCode (`~/.config/opencode/skills/`).

Consider supporting both `~/.gemini/antigravity/skills` and `~/.gemini/config/skills` in `_detect_skills_root()` and `is_agentic_runtime`.

File: references/apigee-skills-serving/bin/provision.sh [hunk 5]
```diff
@@ -0,0 +1,807 @@
...
+}
+
+while [ "$#" -gt 0 ]; do
+  case "$1" in
+    --project)                    PROJECT="${2:-}"; shift 2 ;;
+    --location)                   LOCATION="${2:-}"; shift 2 ;;
+    --bucket)                     BUCKET="${2:-}"; shift 2 ;;
+    --apigee-org)                 APIGEE_ORG="${2:-}"; shift 2 ;;
+    --key-path)                   KEY_PATH="${2:-}"; shift 2 ;;
+    --runtime)                    RUNTIME="${2:-}"; shift 2 ;;
```
Comment (OmidTahouri) [inline references/apigee-skills-serving/bin/provision.sh]: Since `set -u` is enabled, passing an unrecognized `--runtime` value leaves `SKILLS_ROOT` unset in the `case "$RUNTIME"` block at lines 738-742, causing an unbound variable failure at line 743 (`KEYS_DIR="${SKILLS_ROOT}/skill-finder/keys"`).

Consider validating `--runtime` during argument parsing:
```suggestion
    --runtime)
      RUNTIME="${2:-}"
      case "$RUNTIME" in
        opencode|gemini|antigravity) ;;
        *)
          echo "[provision] FATAL: --runtime must be opencode | gemini | antigravity (got: $RUNTIME)" >&2
          exit 1
          ;;
      esac
      shift 2
      ;;
```

File: references/apigee-skills-serving/skills/skill-finder/keys/README.md [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+  self-signed skills), or
+- pin them to a key someone in this repo can rotate under
+  them (governance nightmare).
+
+Instead: this repo ships the source code and the installer,
+and the trust root is materialised on the customer's machine
+during provisioning. Nothing signed by anybody's key ever
+enters this repo.
+
+See `docs/trust-root.md` for the full trust model discussion.
```
Comment (OmidTahouri) [inline references/apigee-skills-serving/skills/skill-finder/keys/README.md]: `docs/trust-root.md` does not exist; please update the reference to point to `docs/architecture.md#trust-model`.

```suggestion
See `docs/architecture.md#trust-model` for the full trust model discussion.
```

File: references/apigee-skills-serving/skills/skill-publisher/SKILL.md [hunk 7]
```diff
@@ -0,0 +1,312 @@
...
+  touches. It is read by `sign_skill.py`, never logged, never
+  uploaded. The key file should have mode `0600`.
+- The signature is computed over the **canonical** manifest bytes
+  (see `scripts/common/canonical.py`). Any post-signing edit to
+  the manifest invalidates the signature; `skill-finder` will
+  refuse to install the skill.
+- The `signing_key_id` field is the sha256 of the **public** key,
+  not the private key. It is safe to publish. `skill-finder`
+  cross-checks it against its embedded trust root
+  (`keys/trusted_pubkey.pem`); a manifest signed by a key not in
```
Comment (OmidTahouri) [inline references/apigee-skills-serving/skills/skill-publisher/SKILL.md]: This still references the single-key model (`keys/trusted_pubkey.pem`). Please update to reflect the multi-key trust root (`keys/*.pem`).

```suggestion
  cross-checks it against installed trust roots
  (`keys/*.pem`); a manifest signed by a key not in
```

File: references/apigee-skills-serving/bin/install-skill-finder.sh (hunk 1 above)
Comment (gsjurseth) [reply references/apigee-skills-serving/bin/install-skill-finder.sh:440]: Good catch — fixed in 4beed4fb.

Moved a `mkdir -p "$INSTALL_ROOT"` immediately above the `mktemp -d -p` so the staging dir creation no longer depends on the install-phase mkdir that runs later in the script. Verified by running the installer against a HOME with no pre-existing `.local/share/` tree.

File: references/apigee-skills-serving/bin/install-skill-finder.sh (hunk 2 above)
Comment (gsjurseth) [reply references/apigee-skills-serving/bin/install-skill-finder.sh]: Fixed in 4beed4fb — updated to `docs/architecture.md#trust-model` per your suggestion. The `docs/trust-root.md` reference was a leftover from an earlier structure where the trust discussion had its own file; it was merged into architecture.md but the pointer never got updated.

File: references/apigee-skills-serving/skills/skill-publisher/scripts/publish.sh (hunk 3 above)
Comment (gsjurseth) [reply references/apigee-skills-serving/skills/skill-publisher/scripts/publish.sh:101]: Fixed in 4beed4fb, applied your suggested `tr` pipeline verbatim. Also grepped the rest of publish.sh and both installers for other Bash 4-isms (`${var,,}`, `${var^^}`, `declare -A`, `mapfile`, `readarray`) — none remain. The portability claim on line 57 now holds.

File: references/apigee-skills-serving/skills/skill-finder/scripts/find_install.py (hunk 4 above)
Comment (gsjurseth) [reply references/apigee-skills-serving/skills/skill-finder/scripts/find_install.py]: Great find — this was a real bug that would have silently degraded every installer-created antigravity install to the OpenCode reload-hint path. Fixed in 4beed4fb.

Approach:
* Replaced the singular `_ANTIGRAVITY_ROOT` with an `_ANTIGRAVITY_ROOTS` tuple containing both `~/.gemini/antigravity/skills/` (canonical, matches the installer + provision.sh) AND the legacy `~/.gemini/config/skills/`.
* `_detect_skills_root()` now iterates the tuple and returns the *matched* root, so installs under either path stay stable in place.
* Kept a `_ANTIGRAVITY_ROOT = _ANTIGRAVITY_ROOTS[0]` back-compat alias so callers wanting the single canonical path still work.
* `is_agentic_runtime` in `main` now unpacks `*_ANTIGRAVITY_ROOTS`, so the reload-hint trailer fires correctly on both.
* Added `tests/test_find_install_antigravity_paths.py` with 4 regression tests pinning both roots + the alias invariant. Full suite is now 241 passing (up from 237).

File: references/apigee-skills-serving/bin/provision.sh (hunk 5 above)
Comment (gsjurseth) [reply references/apigee-skills-serving/bin/provision.sh]: Fixed in 4beed4fb using your suggested pattern (inline case validation in the `--runtime` arm). Also added a defense-in-depth `*)` arm to the auto-detect SKILLS_ROOT case (line ~757 in the new tree) so any future skew between the two `RUNTIME=` write sites still fails loudly rather than silently unbinding `SKILLS_ROOT`.

File: references/apigee-skills-serving/skills/skill-finder/keys/README.md (hunk 6 above)
Comment (gsjurseth) [reply references/apigee-skills-serving/skills/skill-finder/keys/README.md]: Fixed in 4beed4fb — updated to `docs/architecture.md#trust-model`. Same root cause as the installer's stale ref.

File: references/apigee-skills-serving/skills/skill-publisher/SKILL.md (hunk 7 above)
Comment (gsjurseth) [reply references/apigee-skills-serving/skills/skill-publisher/SKILL.md]: Fixed in 4beed4fb — updated to `keys/*.pem` and expanded the sentence slightly to call out that multi-key install supports both multi-publisher orgs and zero-downtime key rotation, matching the language in `docs/architecture.md#trust-model`.

**Comment by gsjurseth** [comment]
@OmidTahouri thank you for the thorough review — all seven inline comments plus the taxonomy-table note are addressed in 4beed4fb.

**Summary of the push:**

| # | Item | Kind | Resolution |
|---|------|------|------------|
| 1 | `install-skill-finder.sh:437` mkdir/mktemp order | bug | added `mkdir -p "$INSTALL_ROOT"` above the `mktemp -d -p` |
| 2 | `install-skill-finder.sh:546` `docs/trust-root.md` | doc | → `docs/architecture.md#trust-model` |
| 3 | `publish.sh:98` Bash 4 `${var,,}` | portability | replaced with `tr '[:upper:]' '[:lower:]'` pipeline |
| 4 | `find_install.py` antigravity path mismatch | bug | `_ANTIGRAVITY_ROOTS` tuple now covers both canonical (`~/.gemini/antigravity/skills`) and legacy (`~/.gemini/config/skills`) layouts in both `_detect_skills_root` and `is_agentic_runtime`; added 4 regression tests |
| 5 | `provision.sh:178` unvalidated `--runtime` | bug | inline case validation at parse time + defense-in-depth `*)` arm in the auto-detect SKILLS_ROOT case |
| 6 | `keys/README.md:51` `docs/trust-root.md` | doc | → `docs/architecture.md#trust-model` |
| 7 | `skill-publisher/SKILL.md:267` single-key ref | doc | → `keys/*.pem` with a note on rotation/multi-publisher |
| — | `docs/architecture.md` taxonomy table (from review body) | doc | rewrote table to match shipped schema (`agentic_skill`, `keywords`, `gs_uri`, `signing_key_id`) and added a row for the system-defined `skill-spec` enum patched from `provision.sh` |

**Verification:**
* `shellcheck` — clean on all four edited bash scripts.
* `bash -n` — clean.
* `pytest tests/` — **241 passed** (previously 237; +4 new antigravity-path regressions).

Ready for another look when you have a moment.

**Review by OmidTahouri — APPROVED** [review]
LGTM
